### PR TITLE
google_iam_binding_uses_role

### PR DESCRIPTION
**Pull request by mknoedel**
Direct:
image

Mapped: 
image

### CODE REVIEW

File: src/steps/cloud-asset/index.ts [hunk 1]
```diff
@@ -48,10 +60,48 @@ export async function fetchIamBindings(
...
+                  IAM_ROLE_ENTITY_TYPE,
+                ),
+                properties: {
+                  projectId,
+                },
+                _mapping: {
+                  relationshipDirection: RelationshipDirection.FORWARD,
+                  sourceEntityKey: bindingEntity._key,
+                  targetFilterKeys: [['_type', '_key']],
+                  skipTargetCreation: false,
```
Comment (mknoedel) [inline src/steps/cloud-asset/index.ts:94]: Do we want to skip target creation here? from the looks of it, the vast majority of iam_roles are not ingested by this integration. I think they are google managed or something like that.

File: src/steps/cloud-asset/index.ts [hunk 2]
```diff
@@ -48,10 +60,48 @@ export async function fetchIamBindings(
           }
         }
 
-        await jobState.addEntity(
+        const bindingEntity = await jobState.addEntity(
           createIamBindingEntity({ _key, projectId, binding, resource }),
         );
 
+        if (binding.role) {
```
Comment (mknoedel) [inline src/steps/cloud-asset/index.ts:67]: Do we need to account for project-level role names differently (ex: `projects/j1-gc-integration-dev-v3/roles/167984947943customrole`)? Not that I could tell from our test environment, but it was something that stood out as other bindings looked to be `groups/admin` and stuff like that.

File: src/steps/cloud-asset/index.ts [hunk 3]
```diff
@@ -48,10 +60,48 @@ export async function fetchIamBindings(
...
+            await jobState.addRelationship(
+              createMappedRelationship({
+                _class: RelationshipClass.USES,
+                _type: generateRelationshipType(
+                  RelationshipClass.USES,
+                  bindingEntities.BINDINGS._type,
+                  IAM_ROLE_ENTITY_TYPE,
+                ),
+                properties: {
+                  projectId,
```
Comment (austinkelleher) [inline src/steps/cloud-asset/index.ts:88]: We should probably remove this. I think that this role could be a custom role that exists as part of a child project that will get picked up later by another integration instance.

File: src/steps/cloud-asset/index.ts (hunk 1 above)
Comment (austinkelleher) [reply src/steps/cloud-asset/index.ts:94]: Similar to above, I think we should create this target entity. This is likely a "custom" role owned by another child project integration instance.

File: src/steps/cloud-asset/index.ts [hunk 4]
```diff
@@ -48,10 +60,48 @@ export async function fetchIamBindings(
...
-        await jobState.addEntity(
+        const bindingEntity = await jobState.addEntity(
           createIamBindingEntity({ _key, projectId, binding, resource }),
         );
 
+        if (binding.role) {
+          const roleEntity = await jobState.findEntity(binding.role);
+
+          if (roleEntity) {
+            await jobState.addRelationship(
```
Comment (austinkelleher) [inline src/steps/cloud-asset/index.ts:71]: Please make sure that we capture direct relationships to "managed" role entities.

**Review by austinkelleher — CHANGES_REQUESTED** [review]

File: src/steps/cloud-asset/index.ts (hunk 4 above)
Comment (mknoedel) [reply src/steps/cloud-asset/index.ts:71]: Done by pulling this out into it's own step and then making it dependent on the role-fetching and creating steps.

**Review by austinkelleher — APPROVED** [review]
Amazing stuff! Super excited about these changes!

**Pull request merged**